examples/pna-example-tcp-connection-tracking.p4

```diff
@@ -229,18 +231,39 @@ control MainControlImpl(
     }
 
     action ct_tcp_table_miss() {
+#ifdef AVOID_IF_INSIDE_ACTION
+        // Note that this code is NOT completely equivalent to the
+        // code in the #else branch below, because it does not call
+        // drop_packet() if do_add_on_miss is false.  One way to make
+        // the two versions of the program equivalent, not shown here,
+        // would be to assign a value to a metadata variable here such
+        // as 'ct_tcp_table_got_miss = true;', and then write
+        // additional code after ct_tcp_table.apply() equivalent in
+        // behavior to this:
+        //
+        // if (ct_tcp_table_got_miss && !do_add_on_miss) {
+        //     drop_packet();
+        // }
+        add_status =
+            add_entry_if(do_add_entry = do_add_on_miss,
```

Comment:
Its not clear where ```do_add_on_miss``` is evaluated ?

Reply:
I do not understand why it is unclear.  `do_add_on_miss` is the expression that is the first parameter of `add_entry_if`.  In P4, for any call to an action, function, extern function, or extern method, expressions that are the parameters to the call are first evaluated, then once their values (or references, if it is an L-value for an `out` or `inout` parameter) are determined, then the call is made.

If that is unclear, please try asking your question in a different way, or feel free to email me at andy.fingerhut@gmail.com and we can schedule time to talk live.

Reply:
@shirshyad Perhaps you might not have seen the syntax in P4 that this example uses, where when you write an action like `action foo(bit<4> a, bit<8> b)` you are allowed to call it with syntax like this `foo(a = 5, b = 7)`.  It is much more common to write `foo(5, 7)`, but both are allowed and have the same effect.  I used the first style of syntax in this example to try to make it clearer what each parameter meant, but may have ended up confusing readers in the process.  We can remove the `parameter_name =` occurrences from this example if you think it would be clearer.

Reply:
I am aware of ```foo(a=5, b=7)``` usage.  At line 248, I see do_add_on_miss. I was searching for ```do_add_on_miss``` as a action local variable, action parameter etc.  Hence the question.

Reply:
do_add_on_miss in that example P4 program is a variable defined inside the control, before the action where it is used.  P4 actions can refer to any variable that is in scope, including local variables declared earlier in the control that they are defined within (if the action is defined within a control), or any parameters of the control, as well as parameters of the action.